keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,582 @@
+
+### The Concept of Node Map and Memory Maps
+
+*Memory Manager* has an internal database, i.e. `Node Map` that incorporates `Memory Maps`. This database is used to keep track of the memory reserved for already deployed containers in Guaranteed QoS class. `Node Map` object remembers the dynamic configuration of disjont NUMA-node groups in `Node` objects. Notice that in fact `Memory Map` offers the counters for tracking the memory; therefore, the map should not be understood as a map that allows for reservation of ranges of memory or consecutive blocks of memory. 
+
+The map is also utilized to calculate NUMA affinity once a container is being deployed. *Memory Manager* supports conventional memory and hugepages of various and all possible sizes (such as 2 MiB or 1 GiB). *Memory Manager* takes all those kinds of memory into account for topology hints calculation and generation. For instance, consider a *kubelet* node which offers hugepages of a couple of sizes, e.g., 1 GiB and 2 MiB. In other words, the node exposes three types of memory to *Memory Manager*, i.e.: conventional memory, `hugepages-1Gi`, and `hugepages-2Mi`.
+
+At the start-up time, *Memory Manager* initializes a `Memory Table` collection for each NUMA node and respective memory types resulting in ready-to-be-used `Memory Map` objects. The code below illustrates the way both `Memory Table` and `Memory Map` collections are structured.
+
+```go
+// MemoryTable contains memory information
+type MemoryTable struct {
+	TotalMemSize   uint64 `json:"total"`
+	SystemReserved uint64 `json:"systemReserved"`
+	Allocatable    uint64 `json:"allocatable"`
+	Reserved       uint64 `json:"reserved"`
+	Free           uint64 `json:"free"`
+}
+
+// NodeState contains NUMA node related information
+type NodeState struct {
+	// NumberOfAssignments contains a number memory assignments from this node
+	// When the container requires memory and hugepages it will increase number of assignments by two
+	NumberOfAssignments int `json:"numberOfAssignments"`
+	// MemoryTable contains NUMA node memory related information
+	MemoryMap map[v1.ResourceName]*MemoryTable `json:"memoryMap"`
+	// NodeGroups contains NUMA nodes that current NUMA node in group with them
+	// It means that we have container that pinned to the current NUMA node and all group nodes
+	Nodes []int `json:"nodes"`
+}
+
+// NodeMap contains memory information for each NUMA node.
+type NodeMap map[int]*NodeState
+```
+
+For instance, consider a two-socket machine with only one size of hugepages pre-allocated. In this case, *Memory Manager* instatiates four memory tables (`[ two NUMA nodes ] * [ two  memory types ]`) on start-up. Notice that on a single-socket machine, and in the case when sub-NUMA cluster (SNC) feature is enabled in hardware, *Memory Manager* can detect two NUMA nodes although, as already mentioned, the machine features a single socket.
```

Comment:
I thought that with sub-NUMA clustering the cores/cache in a socket are split in two, maybe with each cluster getting its own memory controller.  So any given core is only associated with a single NUMA node at at time.

As you say though it's only one of a number of things that might cause there to be multiple NUMA nodes on a single socket.  It might make sense to reword the last sentence as something like:

"Notice that on a single-socket machine there are scenarios (such as the case when sub-NUMA cluster (SNC) feature is enabled in hardware) where *Memory Manager* can detect two NUMA nodes although the machine features a single socket."

Reply:
From [1]:

> Cluster-on-die (COD) and sub-NUMA-Cluster (SNC) are technologies where a physical NUMA node is further split into separate NUMA domains, taking advantage of multiple memory controllers per physical CPU socket. COD and SNC improve the performance of highly NUMA optimized workloads.

> These settings are not directly detectable from Linux user space. However, we can detect the number of NUMA nodes and number of physical IDs of CPUs. If the number of NUMA nodes is greater than the number of physical IDs, it is likely that COD/SNC is enabled. There are corner cases like numa emulation (which isn't likely be used in a k8s cluster).

From [3]:
> What SNC does is divide the silicon at a firmware level into two ‘clusters’, with each cluster having a preference for working with the cores, nodes, and memory controllers within its own cluster. There is nothing stopping it going outside its own cluster, but to offer the best latency (sometimes at the expense of peak bandwidth), it is best for each core/node to be limited in this way.

From [2], an example of latency in a two-socket machine with SNC enabled (from the lowest latency to the highest):
1. "within one zone is ~68ns" (CPU access the most local memory, I guess)
2. "Within one die ~75ns"
3. "Between dies, it is ~125ns, almost eliminating difference between zones within remote die"
4. "Going to another socket ~135ns (again, laws of physics, the distance of signal travelling)."

It seems that it's all about distances, i.e.: shorter distance => lower latency => higher performance (more effective utilization of CPU since waiting for memory is shorter).

It was reworded to:
> Notice that on modern architectures, a single socket, i.e., a physical NUMA node, can be further split into a number of logical NUMA nodes, which is done to further improve the performance. For this purpose, Intel offers machines with Sub-NUMA Clustering (SNC, an upgraded version of Cluster-On-Die) feature, and AMD offers mchines with NUMA Nodes per Socket (NPS) feature. In such a scenario where SNC or NPS is enabled on a machine, a K8S node and the *Memory Manager* can detect and be aware of many NUMA nodes, although the machine features a single socket (a single physical NUMA node).

@klueska @cbf123 Can you recall other scenarios beside SNC/NPS and "numa emulation (which isn't likely be used in a k8s cluster)"? (not crucial to know just can be interesting)
  

[1] https://github.com/kubernetes-sigs/node-feature-discovery/issues/143
[2] slides 10-11, https://docs.google.com/presentation/d/14Hk7uAuhpIACrUmPXFwedwy2TRzdvPJt6OQVNy_WHGE/edit?usp=sharing
[3] https://www.anandtech.com/show/12409/intel-launches-xeon-d-2100-series-socs-edge/4